soc/xtensa/esp32/soc.c

```diff
@@ -84,3 +107,75 @@ int arch_printk_char_out(int c)
 	esp32_rom_uart_tx_one_char(c);
 	return 0;
 }
+
+void sys_arch_reboot(int type)
+{
+	esp_restart_noos();
+}
+
+void IRAM_ATTR esp_restart_noos(void)
+{
+	/* Disable interrupts */
+	z_xt_ints_off(0xFFFFFFFF);
+
+	const uint32_t core_id = cpu_hal_get_core_id();
+	const uint32_t other_core_id = (core_id == 0) ? 1 : 0;
+
+	soc_ll_reset_core(other_core_id);
+	soc_ll_stall_core(other_core_id);
+
+	/* Flush any data left in UART FIFOs */
+	esp32_rom_uart_tx_wait_idle(0);
+	esp32_rom_uart_tx_wait_idle(1);
+	esp32_rom_uart_tx_wait_idle(2);
```

Comment:
What is the reason to flush UART before doing reset?

Reply:
This is part of esp-idf implementation. Perhaps a way to avoid uart output trash after reset.